src/Symfony/Component/Stream/FilesystemStreamWrapper.php

```diff
@@ -0,0 +1,188 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Stream;
+
+use Symfony\Component\Filesystem\Filesystem;
+
+/**
+ * Native filesystem stream wrapper.
+ *
+ * @author Roland Franssen <franssen.roland@gmail.com>
+ *
+ * @internal
+ */
+class FilesystemStreamWrapper implements StreamWrapperInterface
+{
+    /** @var resource|null */
+    public $context;
+
+    private $baseDir;
+    private $filesystem;
+    private $directory;
+
+    public function __construct(string $baseDir, Filesystem $filesystem = null)
+    {
+        $this->baseDir = ltrim($baseDir, '/\\');
+        $this->filesystem = $filesystem ?? new Filesystem();
+    }
+
+    public function dir_closedir(): bool
+    {
+        if (null === $this->directory) {
+            throw new \RuntimeException('Directory resource not available.');
+        }
+
+        closedir($this->directory);
```

Comment:
e.g. closedir() is allowed to be called without a resource, in case it'll close the last opened directory.

here we assert for a resource above. my reasoning is that if there's a "last directory" to be closed, it would be set here in $this->directory already. But im not sure it's OK in practice, e.g. Drupal doesnt do it: https://github.com/drupal/drupal/blob/8.6.x/core/lib/Drupal/Core/StreamWrapper/LocalStream.php#L565

but calling closedir with null doesnt work.. so who knows :) https://3v4l.org/GSI6V

This should be fed by test scenarios i guess.